src/management/Akka.Http.Shim/Dsl/Model/HttpMessage.cs

```diff
@@ -0,0 +1,166 @@
+using System;
+using System.Collections.Immutable;
+using System.Net;
+using Akka.Annotations;
+using Akka.IO;
+using Microsoft.AspNetCore.Http;
+
+namespace Akka.Http.Dsl.Model
+{
+    /// <summary>
+    /// The base type for an Http message (request or response).
+    /// </summary>
+    [DoNotInherit]
+    public abstract class HttpMessage<T>
+    {
+        /// <summary>
+        /// The entity of this message.
+        /// </summary>
+        public abstract ResponseEntity Entity { get; }
+
+        /*
+        /// <summary>
+        /// Returns a copy of this message with the entity set to the given one.
+        /// </summary>
+        public abstract T WithEntity(RequestEntity entity);
+
+        public T WithEntity(string content) =>
+            WithEntity((RequestEntity)HttpEntity.Create(content));
+
+        public T WithEntity(ByteString bytes) =>
+            WithEntity((RequestEntity)HttpEntity.Create(bytes));
+
+        public T WithEntity(string contentType, string content) =>
+            WithEntity((RequestEntity)HttpEntity.Create(contentType, content));
+        */
+    }
+
+    /// <summary>
+    /// The immutable HTTP request model.
+    /// </summary>
+    public sealed class HttpRequest : HttpMessage<HttpRequest>
+    {
+        /// <summary>
+        /// Returns the Http method of this request.
+        /// </summary>
+        public string Method => _request.Method;
+
+        /// <summary>
+        /// Returns the Uri of this request.
+        /// </summary>
+        public PathString Path => _request.Path;
+
+        public IPAddress Peer => _request.HttpContext.Connection.RemoteIpAddress;
+
+        /// <summary>
+        /// Returns the entity of this request.
+        /// </summary>
+        public override ResponseEntity Entity { get; }
+
+        private readonly Microsoft.AspNetCore.Http.HttpRequest _request;
+
+        public static HttpRequest Create(Microsoft.AspNetCore.Http.HttpRequest request) =>
+            new HttpRequest(request);
+
+        private HttpRequest(Microsoft.AspNetCore.Http.HttpRequest request)
+        {
+            _request = request;
+            var input = new byte[Convert.ToInt32(request.ContentLength)];
+            request.Body.Read(input, 0, input.Length);
+
+            Entity = new RequestEntity(request.ContentType, ByteString.FromBytes(input));
+        }
```

Comment:
Do we really want this? I was trying to avoid leaking anything from AspNetCore outside the Http extesion itself. By leveraging the [Request Features](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/request-features) via `HttpRequestMiddleware`, we can do these mappings at the middleware level.

Is there any limitations to that approach?

Reply:
Not really, I was just trying to reduce the complexity of the shim. Except for the possibility that Microsoft changes the API (yet again) is there a danger of doing it this way? Is there really a need to completely decouple our code from ASP.NET?

Reply:
Referenced this in a new issue: https://github.com/akkadotnet/Akka.Management/issues/17

Still think we should merge this and push a new beta package so we can start testing out some the functionality, with the expectation that one way or another the Akka.HTTP APIs will evolve from here (i.e. they are not yet frozen)